File: backend/schemas/rating.py
from pydantic import BaseModel, ConfigDict, Field

from backend.schemas.comment import CommentResponse, CommunityAuthor


class RatingSubmitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    score: float = Field(ge=0, le=10, multiple_of=0.1)
    reason: str | None = Field(default=None, max_length=100)


class RatingSubmitResponse(BaseModel):
    rating_id: str
    score: float
    reason: str | None
    average: float
    rating_count: int


class RatingReasonResponse(BaseModel):
    id: str
    author: CommunityAuthor
    score: float
    reason: str
    like_count: int
    liked_by_me: bool
    updated_at: str


class ViewerPermissions(BaseModel):
    authenticated: bool
    is_supersonic_player: bool
    can_rate: bool
    can_comment: bool
    can_like: bool


class RatingMatchTeam(BaseModel):
    id: str
    name: str
    crest_url: str | None = None


class RatingMatchSummary(BaseModel):
    id: str
    season_id: str
    competition: str
    round: int | None = None
    date: str | None = None
    home_team: RatingMatchTeam
    away_team: RatingMatchTeam
    home_score: int
    away_score: int


class MatchFacts(BaseModel):
    goals: int = 0
    assists: int = 0


class PlayerRatingAggregate(BaseModel):
    average: float | None = None
    count: int = 0
    my_score: float | None = None
    my_reason: str | None = None
    reasons: list[RatingReasonResponse] = Field(default_factory=list)


class RatingPagePlayer(BaseModel):
    player_id: str
    name: str
    number: int | None = None
    photo_url: str | None = None
    match_facts: MatchFacts
    rating: PlayerRatingAggregate
    top_comment: CommentResponse | None = None
    comment_count: int = 0


class FanMvpResponse(BaseModel):
    player_id: str
    name: str
    photo_url: str | None = None
    average: float
    rating_count: int


class RatingMatchPageResponse(BaseModel):
    match: RatingMatchSummary
    viewer: ViewerPermissions
    fan_mvp: FanMvpResponse | None = None
    players: list[RatingPagePlayer] = Field(default_factory=list)

File: frontend/src/types/rating.ts
import type { Team } from './index'

export interface CurrentPlayerUser {
  id: string
  playerId: string
  name: string
  photoUrl: string | null
}

export interface PlayerSession {
  authenticated: true
  user: CurrentPlayerUser
  isSupersonicPlayer: boolean
}

export interface ViewerPermissions {
  authenticated: boolean
  isSupersonicPlayer: boolean
  canRate: boolean
  canComment: boolean
  canLike: boolean
}

export interface CommunityAuthor {
  playerId: string
  name: string
  photoUrl: string | null
}

export interface CommunityComment {
  id: string
  author: CommunityAuthor
  content: string
  likeCount: number
  likedByMe: boolean
  createdAt: string
  updatedAt: string
}

export interface RatingReason {
  id: string
  author: CommunityAuthor
  score: number
  reason: string
  likeCount: number
  likedByMe: boolean
  updatedAt: string
}

export interface RatingSummary {
  average: number | null
  count: number
  myScore: number | null
  myReason: string | null
  reasons: RatingReason[]
}

export interface RatingPlayer {
  playerId: string
  name: string
  number: number | null
  photoUrl: string | null
  matchFacts: { goals: number; assists: number }
  rating: RatingSummary
  topComment: CommunityComment | null
  commentCount: number
}

export interface RatingMatchSummary {
  id: string
  seasonId: string
  competition: string
  round: number | null
  date: string | null
  homeTeam: Team
  awayTeam: Team
  homeScore: number
  awayScore: number
}

export interface FanMvp {
  playerId: string
  name: string
  photoUrl: string | null
  average: number
  ratingCount: number
}

export interface RatingMatchResponse {
  match: RatingMatchSummary
  viewer: ViewerPermissions
  fanMvp: FanMvp | null
  players: RatingPlayer[]
}
